baseline/baseline.py

```diff
@@ -82,79 +105,169 @@ def load_model(model_file_path: str, device: str = "cpu") -> Any:
     return RandomProjectionMelEmbedding().to(device)
 
 
+def frame_audio(audio: Tensor, frame_size: int, hop_size: int, is_centered: bool):
+    """
+    Slice audio into equal length frames. Each adjacent frame is a hop_size number of
+    samples apart.
+
+    Args:
+        audio: input audio, expects a 2d Tensor of shape: (batch_size, num)samples)
+        frame_size: the length each resulting frame should be
+        hop_size: number of samples between frames.
+        is_centered: Pad audio by frame_size // 2 to center audio in frame
+
+    Returns:
+        A Tensor of shape (batch_size, num_frames, frame_size)
+    """
+    batch_size, num_samples = audio.shape
+
+    # Adjust the number of samples if centering to allow for a half
+    # frame size number of padded samples padded at the start.
+    start_pad = 0
+    if is_centered:
+        half_frame_size = int(frame_size // 2)
+        num_samples += half_frame_size
+        start_pad = half_frame_size
+
+    # Number of frames is the number of hops that can occur within num_samples
+    num_frames = math.ceil(num_samples / hop_size)
+
+    # Pad audio to facilitate centered frames.
+    padded_num_samples = (num_frames - 1) * hop_size + frame_size
+    end_pad = padded_num_samples - num_samples
+    audio = F.pad(audio, (start_pad, end_pad))
+
+    # Frame.
+    shape = (batch_size, num_frames, frame_size)
+    stride = (1, hop_size, 1)
+    frames = torch.as_strided(audio, shape, stride)
+
+    return frames
+
+
+def get_timestamps_for_embedding(
+    sample_rate: int,
+    num_frames: int,
+    hop_size: int,
+    frame_size: Optional[int] = None,
+    center: Optional[bool] = True,
+) -> Tensor:
+    """
+    Returns a tensor of timestamps in seconds that correspond to the time
+    locations of each embedding.
+
+    Args:
+        sample_rate: audio sampling rate of input audio
+        num_frames: number of frames to compute timestamps for
+        hop_size: distance between adjacent frames used to calculate
+        frame_size: used to calculate time offset when center is False
+        center: whether of not the frames were padded to center audio, if false
+            an offset will be applied equal to frame_size // 2.
+
+    Returns:
+        Timestamps in seconds.
+    """
+
+    offset = 0
+    if not center:
+        if frame_size is None:
```

Comment:
I don't think there's a need for this. Timestamps will be the same whether or not the frames are centered.

Reply:
I thought that we always calculated the timestamp for a frame as being the center of the frame -- at least for spectral stuff? Was referring to https://librosa.org/doc/latest/generated/librosa.frames_to_samples.html#librosa.frames_to_samples -- (which is used to calculate timestamps for frames).

Reply:
So if there isn't centering / zero-padding then the first timestamp will be frame_size // 2